File: seed-data.js
/**
 * Pixel Mosaic Maker — sample data seed
 *
 * Populates all six galleries with a small set of demo items so you can
 * explore the full workflow without uploading your own images.
 *
 * Usage (browser dev tools console, on any page of the running app):
 *   copy the contents of this file and paste into the console, then press Enter.
 *   Reload the page after running it.
 *
 * Two demos are included:
 *   • Smiley Face   — 32×32 pixel-art face, single-panel layout, step size 16/32
 *   • Quilt Pattern — 32×32 geometric quilt, 2×2 layout, step size 16
 */
(function () {

  // ---------------------------------------------------------------------------
  // Helpers
  // ---------------------------------------------------------------------------

  function countColors(pixels) {
    const counts = {};
    pixels.forEach(function (p) { counts[p] = (counts[p] || 0) + 1; });
    return counts;
  }

  // ---------------------------------------------------------------------------
  // Pixel art generators
  // ---------------------------------------------------------------------------

  /**
   * 32×32 smiley face.
   * Palette indices:
   *   0 = white background
   *   1 = golden yellow face
   *   2 = dark brown (eyes & mouth)
   *   3 = pink (cheeks)
   */
  function makeSmilePixels() {
    var W = 32, H = 32;
    var cx = 15.5, cy = 15.5, r2 = 12.5 * 12.5;
    var pixels = [];
    for (var y = 0; y < H; y++) {
      for (var x = 0; x < W; x++) {
        var dx = x - cx, dy = y - cy;
        var inFace = (dx * dx + dy * dy) <= r2;

        // Eyes — two 2×2 dark blocks
        var leftEye  = (x >= 10 && x <= 11) && (y >= 11 && y <= 12);
        var rightEye = (x >= 20 && x <= 21) && (y >= 11 && y <= 12);

        // Cheeks — two 4×2 pink blocks (clipped to face circle)
        var leftCheek  = inFace && (x >= 6  && x <= 9)  && (y >= 17 && y <= 18);
        var rightCheek = inFace && (x >= 22 && x <= 25) && (y >= 17 && y <= 18);

        // Smile — U-shaped dark curve
        var smile = inFace && (
          ((x === 9  || x === 10) && (y >= 20 && y <= 21)) ||
          ((x === 21 || x === 22) && (y >= 20 && y <= 21)) ||
          ((x >= 11 && x <= 20)   && (y >= 22 && y <= 23))
        );

        if (!inFace)                       pixels.push(0);
        else if (leftEye || rightEye)      pixels.push(2);
        else if (smile)                    pixels.push(2);
        else if (leftCheek || rightCheek)  pixels.push(3);
        else                               pixels.push(1);
      }
    }
    return pixels;
  }

  /**
   * 32×32 quilt: 4×4 tiles of 8×8px each, rotated color palette per tile.
   * A 1px white grid line separates every tile (x%8===0 or y%8===0).
   * Palette indices:
   *   0 = white (grid lines)
   *   1 = red
   *   2 = blue
   *   3 = green
   *   4 = yellow/gold
   */
  function makeQuiltPixels() {
    var tc = [[1,2,3,4],[2,3,4,1],[3,4,1,2],[4,1,2,3]];
    var pixels = [];
    for (var y = 0; y < 32; y++) {
      for (var x = 0; x < 32; x++) {
        if (x % 8 === 0 || y % 8 === 0) {
          pixels.push(0);
        } else {
          pixels.push(tc[Math.floor(y / 8)][Math.floor(x / 8)]);
        }
      }
    }
    return pixels;
  }

  // ---------------------------------------------------------------------------
  // Build data
  // ---------------------------------------------------------------------------

  var smilePixels = makeSmilePixels();
  var quiltPixels = makeQuiltPixels();

  // ---- Palettes ----

  var PALETTES = [
    {
      id: 'sample-palette-smiley',
      name: 'Sample: Smiley Colors',
      colors: [
        { r: 255, g: 255, b: 255 }, // 0 → background (white)
        { r: 255, g: 210, b: 30  }, // 1 → face (golden yellow)
        { r: 45,  g: 25,  b: 5   }, // 2 → features (dark brown)
        { r: 255, g: 150, b: 150 }  // 3 → cheeks (pink)
      ]
    },
    {
      id: 'sample-palette-quilt',
      name: 'Sample: Quilt Colors',
      colors: [
        { r: 255, g: 255, b: 255 }, // 0 → grid lines (white)
        { r: 200, g: 45,  b: 45  }, // 1 → red
        { r: 45,  g: 95,  b: 200 }, // 2 → blue
        { r: 45,  g: 165, b: 95  }, // 3 → green
        { r: 255, g: 200, b: 0   }  // 4 → yellow/gold
      ]
    }
  ];

  // ---- Images ----

  var IMAGES = [
    {
      id: 'sample-image-smiley',
      name: 'Sample: Smiley Face',
      pixelPic: {
        width:         32,
        height:        32,
        paletteLookup: [
          { r: 255, g: 255, b: 255, a: 255 },
          { r: 255, g: 210, b: 30,  a: 255 },
          { r: 45,  g: 25,  b: 5,   a: 255 },
          { r: 255, g: 150, b: 150, a: 255 }
        ],
        pixels:  smilePixels,
        palette: countColors(smilePixels),
        name:    'Sample: Smiley Face'
      }
    },
    {
      id: 'sample-image-quilt',
      name: 'Sample: Quilt Pattern',
      pixelPic: {
        width:         32,
        height:        32,
        paletteLookup: [
          { r: 255, g: 255, b: 255, a: 255 },
          { r: 200, g: 45,  b: 45,  a: 255 },
          { r: 45,  g: 95,  b: 200, a: 255 },
          { r: 45,  g: 165, b: 95,  a: 255 },
          { r: 255, g: 200, b: 0,   a: 255 }
        ],
        pixels:  quiltPixels,
        palette: countColors(quiltPixels),
        name:    'Sample: Quilt Pattern'
      }
    }
  ];

  // ---- Layouts ----
  // "Single Panel" — whole 32×32 image as one section (step sizes 16, 32 available)
  // "2×2 Grid"     — four 16×16 quadrants (step size 16 available)

  var LAYOUT_SINGLE = { cols: 1, rows: 1, parts: [
    { x: 0, y: 0, width: 32, height: 32 }
  ]};

  var LAYOUT_2X2 = { cols: 2, rows: 2, parts: [
    { x: 0,  y: 0,  width: 16, height: 16 },
    { x: 16, y: 0,  width: 16, height: 16 },
    { x: 0,  y: 16, width: 16, height: 16 },
    { x: 16, y: 16, width: 16, height: 16 }
  ]};

  var LAYOUTS = [
    { id: 'sample-layout-single', name: 'Sample: Single Panel', config: LAYOUT_SINGLE },
    { id: 'sample-layout-2x2',   name: 'Sample: 2×2 Grid',     config: LAYOUT_2X2   }
  ];

  // ---- Build Configs ----

  var BUILD_CONFIGS = [
    {
      id:       'sample-buildconfig-smiley',
      name:     'Sample: Smiley Mosaic',
      config: {
        grid:      LAYOUT_SINGLE,
        imageRef:  'sample-image-smiley',
        paletteRef:'sample-palette-smiley',
        offsetX:   0,
        offsetY:   0
      }
    },
    {
      id:       'sample-buildconfig-quilt',
      name:     'Sample: Quilt Mosaic',
      config: {
        grid:      LAYOUT_2X2,
        imageRef:  'sample-image-quilt',
        paletteRef:'sample-palette-quilt',
        offsetX:   0,
        offsetY:   0
      }
    }
  ];

  // ---- Builds ----

  var BUILDS = [
    { id: 'sample-build-smiley', name: 'Sample: Smiley Build', buildConfigRef: 'sample-buildconfig-smiley' },
    { id: 'sample-build-quilt',  name: 'Sample: Quilt Build',  buildConfigRef: 'sample-buildconfig-quilt'  }
  ];

  // ---- Print Configs ----

  var PRINT_CONFIGS = [
    {
      id:                     'sample-printconfig-quilt',
      name:                   'Sample: Quilt Print',
      selectedBuildId:        'sample-build-quilt',
      selectedBuildConfigId:  'sample-buildconfig-quilt',
      title:                  'My First Quilt Mosaic',
      stepSizePx:             16,
      pageBackgroundColorHex: '#fdfbe6',
      patchBackgroundColorHex:'#dcdcdc',
      stacked:                true,
      printerMarginMm:        3,
      contentTopOffsetMm:     2
    }
  ];

  // ---------------------------------------------------------------------------
  // Write to localStorage (prepend; skip items whose id already exists)
  // ---------------------------------------------------------------------------

  function seed(key, items) {
    var existing = JSON.parse(localStorage.getItem(key) || '[]');
    var existingIds = {};
    existing.forEach(function (x) { existingIds[x.id] = true; });
    var toAdd = items.filter(function (x) { return !existingIds[x.id]; });
    localStorage.setItem(key, JSON.stringify(toAdd.concat(existing)));
    console.log('Seeded ' + toAdd.length + ' item(s) → ' + key);
  }

  seed('palettes',     PALETTES);
  seed('images',       IMAGES);
  seed('gridConfigs',  LAYOUTS);
  seed('buildConfigs', BUILD_CONFIGS);
  seed('builds',       BUILDS);
  seed('printConfigs', PRINT_CONFIGS);

  console.log('✓ Sample data loaded. Reload the page to see all galleries populated.');

}());

File: src/main/scala/clemniem/screens/SeedData.scala
package clemniem.screens

import cats.effect.IO
import clemniem.*
import io.circe.{Decoder, Encoder}
import io.circe.syntax.*
import org.scalajs.dom

/** Sample data for all six galleries. Called from the About screen "Load sample data" button. */
object SeedData {

  /** Write sample data to localStorage. Returns the number of new items added (skips existing IDs). */
  def seed(): IO[Int] = IO.delay {
    List(
      seedKey[StoredPalette](StorageKeys.palettes, palettes),
      seedKey[StoredImage](StorageKeys.images, images),
      seedKey[StoredLayout](StorageKeys.layouts, layouts),
      seedKey[StoredBuildConfig](StorageKeys.buildConfigs, buildConfigs),
      seedKey[StoredBuild](StorageKeys.builds, builds),
      seedKey[StoredPrintConfig](StorageKeys.printConfigs, printConfigs)
    ).sum
  }

  private def seedKey[A <: StoredEntity](key: String, items: List[A])(using Encoder[List[A]], Decoder[List[A]]): Int = {
    val existing = Option(dom.window.localStorage.getItem(key))
      .flatMap(json => io.circe.parser.decode[List[A]](json).toOption)
      .getOrElse(Nil)
    val existingIds = existing.map(_.id).toSet
    val toAdd       = items.filterNot(a => existingIds.contains(a.id))
    if (toAdd.nonEmpty) {
      val merged = toAdd ++ existing
      dom.window.localStorage.setItem(key, merged.asJson.noSpacesSortKeys)
    }
    toAdd.size
  }

  // ---------------------------------------------------------------------------
  // Pixel art generators
  // ---------------------------------------------------------------------------

  /** 32x32 smiley face. Palette: 0=white, 1=yellow, 2=brown, 3=pink. */
  private def makeSmilePixels(): Vector[Int] = {
    val cx = 15.5; val cy = 15.5; val r2 = 12.5 * 12.5
    (for {
      y <- 0 until 32
      x <- 0 until 32
    } yield {
      val dx         = x - cx; val dy = y - cy
      val inFace     = (dx * dx + dy * dy) <= r2
      val leftEye    = (x >= 10 && x <= 11) && (y >= 11 && y <= 12)
      val rightEye   = (x >= 20 && x <= 21) && (y >= 11 && y <= 12)
      val leftCheek  = inFace && (x >= 6 && x <= 9) && (y >= 17 && y <= 18)
      val rightCheek = inFace && (x >= 22 && x <= 25) && (y >= 17 && y <= 18)
      val smile      = inFace && (((x == 9 || x == 10) && (y >= 20 && y <= 21)) ||
        ((x == 21 || x == 22) && (y >= 20 && y <= 21)) ||
        ((x >= 11 && x <= 20) && (y >= 22 && y <= 23)))
      if (!inFace) 0
      else if (leftEye || rightEye) 2
      else if (smile) 2
      else if (leftCheek || rightCheek) 3
      else 1
    }).toVector
  }

  /** 32x32 quilt pattern. Palette: 0=white, 1=red, 2=blue, 3=green, 4=gold. */
  private def makeQuiltPixels(): Vector[Int] = {
    val tc = Vector(Vector(1, 2, 3, 4), Vector(2, 3, 4, 1), Vector(3, 4, 1, 2), Vector(4, 1, 2, 3))
    (
      for {
        y <- 0 until 32
        x <- 0 until 32
      } yield if (x % 8 == 0 || y % 8 == 0) 0 else tc(y / 8)(x / 8)
    ).toVector
  }

  private def mkPic(w: Int, h: Int, palette: Vector[Pixel], pixels: Vector[Int], name: String): Option[PixelPic] = {
    val counts = pixels.groupBy(identity).view.mapValues(_.size).toMap
    PixelPic(w, h, palette, pixels, counts, name)
  }

  // ---------------------------------------------------------------------------
  // Sample entities
  // ---------------------------------------------------------------------------

  private val smileyPalette = Vector(
    Pixel(255, 255, 255, 255), // 0 background
    Pixel(255, 210, 30, 255),  // 1 face
    Pixel(45, 25, 5, 255),     // 2 features
    Pixel(255, 150, 150, 255)  // 3 cheeks
  )

  private val quiltPalette = Vector(
    Pixel(255, 255, 255, 255), // 0 grid lines
    Pixel(200, 45, 45, 255),   // 1 red
    Pixel(45, 95, 200, 255),   // 2 blue
    Pixel(45, 165, 95, 255),   // 3 green
    Pixel(255, 200, 0, 255)    // 4 gold
  )

  private lazy val smileyPic: Option[PixelPic] = mkPic(32, 32, smileyPalette, makeSmilePixels(), "Sample: Smiley Face")
  private lazy val quiltPic: Option[PixelPic]  = mkPic(32, 32, quiltPalette, makeQuiltPixels(), "Sample: Quilt Pattern")

  private val palettes: List[StoredPalette] = List(
    StoredPalette("sample-palette-smiley", "Sample: Smiley Colors", smileyPalette.map(p => Color(p.r, p.g, p.b))),
    StoredPalette("sample-palette-quilt", "Sample: Quilt Colors", quiltPalette.map(p => Color(p.r, p.g, p.b)))
  )

  private lazy val images: List[StoredImage] = List(
    smileyPic.map(pic => StoredImage("sample-image-smiley", "Sample: Smiley Face", pic)),
    quiltPic.map(pic => StoredImage("sample-image-quilt", "Sample: Quilt Pattern", pic))
  ).flatten

  private val singlePanel: Layout = Layout(1, 1, Array(GridPart(0, 0, 32, 32)))
  private val grid2x2: Layout     = Layout(
    2,
    2,
    Array(GridPart(0, 0, 16, 16), GridPart(16, 0, 16, 16), GridPart(0, 16, 16, 16), GridPart(16, 16, 16, 16)))

  private val layouts: List[StoredLayout] = List(
    StoredLayout("sample-layout-single", "Sample: Single Panel", singlePanel, None, None, None),
    StoredLayout("sample-layout-2x2", "Sample: 2\u00d72 Grid", grid2x2, None, None, None)
  )

  private val buildConfigs: List[StoredBuildConfig] = List(
    StoredBuildConfig(
      "sample-buildconfig-smiley",
      "Sample: Smiley Mosaic",
      BuildConfig(singlePanel, "sample-image-smiley", "sample-palette-smiley", 0, 0),
      None),
    StoredBuildConfig(
      "sample-buildconfig-quilt",
      "Sample: Quilt Mosaic",
      BuildConfig(grid2x2, "sample-image-quilt", "sample-palette-quilt", 0, 0),
      None)
  )

  private val builds: List[StoredBuild] = List(
    StoredBuild("sample-build-smiley", "Sample: Smiley Build", "sample-buildconfig-smiley", None, None),
    StoredBuild("sample-build-quilt", "Sample: Quilt Build", "sample-buildconfig-quilt", None, None)
  )

  private val printConfigs: List[StoredPrintConfig] = List(
    StoredPrintConfig(
      id = "sample-printconfig-quilt",
      name = "Sample: Quilt Print",
      selectedBuildId = Some("sample-build-quilt"),
      selectedBuildConfigId = Some("sample-buildconfig-quilt"),
      title = "My First Quilt Mosaic",
      stepSizePx = 16,
      pageBackgroundColorHex = "#fdfbe6",
      patchBackgroundColorHex = "#dcdcdc",
      stacked = true,
      printerMarginMm = 3.0,
      contentTopOffsetMm = 2.0
    )
  )
}
